Python/_warnings.c

```diff
@@ -977,25 +977,59 @@ warnings_warn_impl(PyObject *module, PyObject *message, PyObject *category,
 static PyObject *
 get_source_line(PyInterpreterState *interp, PyObject *module_globals, int lineno)
 {
-    PyObject *loader;
+    PyObject *loader, *spec_loader;
+    PyObject *spec;
     PyObject *module_name;
     PyObject *get_source;
     PyObject *source;
     PyObject *source_list;
     PyObject *source_line;
 
-    /* Check/get the requisite pieces needed for the loader. */
+    module_name = _PyDict_GetItemWithError(module_globals, &_Py_ID(__name__));
+    if (!module_name) {
+        return NULL;
+    }
+    Py_INCREF(module_name);
+
+    /* Check/get the requisite pieces needed for the loader.  Get both the
+       __spec__.loader and __loader__ attributes and warn if they are not the
+       same.
+    */
+    spec = _PyDict_GetItemWithError(module_globals, &_Py_ID(__spec__));
+    if (spec == NULL) {
+        Py_DECREF(module_name);
+        return NULL;
```

Comment:
Do we want to fail at this point, or do we want to warn that we are falling through to `__loader__`?

Reply:
That's a good point.  After discussion and thinking this through, here are the cases I think we should handle:

```
    /* This is part of the migration away from module attributes and into a
       fully __spec__ future.  See GH#97850

       * If you have a __loader__ and a __spec__.loader but they are not the
         same, in Python 3.12 we issue a DeprecationWarning and fall back to
         __loader__ for backward compatibility.  In Python 3.14, we'll flip
         this case to ignoring __loader__ entirely, without error.

       * If you do not have a __spec__ or __spec__.loader, we also issue a
         DeprecationWarning and fall back to __loader__.  In Python 3.14,
         we'll make this case fail with an AttributeError, and ignore
         __loader__.

       * In Python 3.12 and beyond, if you do not have a __loader__, we don't
         care as long as you still have a __spec__.loader.
    */
```

I'll work on an update.